lightning/src/util/sweep.rs

```diff
@@ -0,0 +1,800 @@
+				};
+
+				if state_lock
+					.outputs
+					.iter()
+					.find(|o| o.descriptor == output_info.descriptor)
+					.is_some()
+				{
+					continue;
+				}
+
+				state_lock.outputs.push(output_info);
+			}
+			spending_tx_opt = self.regenerate_spend_if_necessary(&mut *state_lock);
+			self.persist_state(&*state_lock).unwrap_or_else(|e| {
+				log_error!(self.logger, "Error persisting OutputSweeper: {:?}", e);
+			});
+		}
+
+		if let Some(spending_tx) = spending_tx_opt {
+			self.broadcaster.broadcast_transactions(&[&spending_tx]);
+		}
+	}
+
+	/// Returns a list of the currently tracked spendable outputs.
+	pub fn tracked_spendable_outputs(&self) -> Vec<TrackedSpendableOutput> {
+		self.sweeper_state.lock().unwrap().outputs.clone()
+	}
+
+	fn regenerate_spend_if_necessary(
+		&self, sweeper_state: &mut SweeperState,
+	) -> Option<Transaction> {
+		let cur_height = sweeper_state.best_block.height;
+		let cur_hash = sweeper_state.best_block.block_hash;
+		let filter_fn = |o: &TrackedSpendableOutput| {
+			if o.status.is_confirmed() {
+				// Don't rebroadcast confirmed txs.
+				return false;
+			}
+
+			if o.status.latest_broadcast_height() >= Some(cur_height) {
+				// Only broadcast once per block height.
+				return false;
+			}
+
+			true
+		};
+
+		let respend_descriptors: Vec<&SpendableOutputDescriptor> =
+			sweeper_state.outputs.iter().filter(|o| filter_fn(*o)).map(|o| &o.descriptor).collect();
+
+		if respend_descriptors.is_empty() {
+			// Nothing to do.
+			return None;
+		}
+
+		let spending_tx = match (self.spend_outputs_callback)(&respend_descriptors) {
+			Ok(spending_tx) => {
+				log_debug!(
+					self.logger,
+					"Generating and broadcasting sweeping transaction {}",
+					spending_tx.txid()
+				);
+				spending_tx
+			},
+			Err(e) => {
+				log_error!(self.logger, "Error spending outputs: {:?}", e);
+				return None;
+			},
+		};
+
+		// As we didn't modify the state so far, the same filter_fn yields the same elements as
+		// above.
+		let respend_outputs = sweeper_state.outputs.iter_mut().filter(|o| filter_fn(&**o));
+		for output_info in respend_outputs {
+			if let Some(filter) = self.chain_data_source.as_ref() {
+				let watched_output = output_info.to_watched_output();
+				filter.register_output(watched_output);
+			}
+
+			output_info.status.broadcast(cur_hash, cur_height, spending_tx.clone());
+		}
+
+		Some(spending_tx)
+	}
+
+	fn prune_confirmed_outputs(&self, sweeper_state: &mut SweeperState) {
+		let cur_height = sweeper_state.best_block.height;
+
+		// Prune all outputs that have sufficient depth by now.
+		sweeper_state.outputs.retain(|o| {
+			if let Some(confirmation_height) = o.status.confirmation_height() {
+				if cur_height >= confirmation_height + ANTI_REORG_DELAY - 1 {
+					log_debug!(self.logger,
+						"Pruning swept output as sufficiently confirmed via spend in transaction {:?}. Pruned descriptor: {:?}",
+						o.status.latest_spending_tx().as_ref().map(|t| t.txid()), o.descriptor
```

Comment:
Is `as_ref` needed here? `latest_spending_tx` already returns an `Option` holding a reference.